### ISSUE TITLE
v2.9.1: Workflow controller restarts with 'fatal error: concurrent map read and map write'

### ISSUE THREAD

**Issue by brabster** [argoproj/argo-workflows#3620]
Checklist:

 * [x] I've included the version.
 * [x] I've included reproduction steps.
 * [x] I've included the workflow YAML.
 * [x] I've included the logs.
 
**What happened**:

Workflow controller started crashing, due to missing end brace in a template.

```
"Observed a panic: &errors.errorString{s:"Cannot find end tag=\"}}\" in the template=\"{\\\"name\\\":\\\"extract-data\\\",\\\"arguments\\\":{},\\\"inputs\\\":{},\\\"outputs\\\":{},\\\"metadata\\\":{},\\\"script\\\":{\\\"name\\\":\\\"\\\",\\\"image\\\":\\\"myimage:latest\\\",\\\"args\\\":[\\\"--main extract\\\",\\\"--triggerDate {{arguments.parameters.triggerDate}\\\"],\\\"resources\\\":{},\\\"volumeMounts\\\":[{\\\"name\\\":\\\"workdir\\\",\\\"mountPath\\\":\\\"/data\\\"}],\\\"source\\\":\\\"\\\"},\\\"serviceAccountName\\\":\\\"my-ksa\\\"}\" starting from \"arguments.parameters.triggerDate}\\\"],\\\"resources\\\":{},\\\"volumeMounts\\\":
```
Earlier error from stacktrace:
```
fatal error: concurrent map read and map write
```

Started up OK when we uninstalled the problematic workflow, before that:
```
NAME                                   READY   STATUS             RESTARTS   AGE
argo-server-6bc845db8-wxdlp            1/1     Running            0          43h
workflow-controller-5cf48bf6b9-f6fh2   0/1     CrashLoopBackOff   14         43h
```

**What you expected to happen**:

Not to crash

**How to reproduce it (as minimally and precisely as possible)**:

Probably missing off a closing brace on an expression but not checked

**Anything else we need to know?**:

No

**Environment**:

- Argo version:
```
$ argo version
argo: vv2.7.4+50b209c.dirty
  BuildDate: 2020-04-16T16:27:43Z
  GitCommit: 50b209ca14c056fb470ebb8329e255304dd5be90
  GitTreeState: dirty
  GitTag: v2.7.4
  GoVersion: go1.13
  Compiler: gc
  Platform: linux/amd64
```
- Kubernetes version :
```
$ kubectl version -o yaml
clientVersion:
  buildDate: "2019-06-19T16:40:16Z"
  compiler: gc
  gitCommit: e8462b5b5dc2584fdcd18e6bcfe9f1e4d970a529
  gitTreeState: clean
  gitVersion: v1.15.0
  goVersion: go1.12.5
  major: "1"
  minor: "15"
  platform: linux/amd64
serverVersion:
  buildDate: "2020-06-01T22:20:10Z"
  compiler: gc
  gitCommit: fb7add51f767aae42655d39972210dc1c5dbd4b3
  gitTreeState: clean
  gitVersion: v1.15.12-gke.2
  goVersion: go1.12.17b4
  major: "1"
  minor: 15+
  platform: linux/amd64
```

**Other debugging information (if applicable)**:

Workflow controller is `argoexec:v2.9.1`
---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

**Comment by jessesuen** [comment]
Could you paste the section of logs which contain the panic stacktrace? At the moment it's not enough to go off of.

**Comment by brabster** [comment]
@alexec yes, it's a GKE cluster.

@jessesuen here you go https://gist.github.com/brabster/4f577e5586164aacaa255ce0df23cff8 - please let me know if you see anything in there I should have redacted, think it's clear!

**Comment by simster7** [comment]
Thanks for the extra informatiion @brabster, I found the issue. Having missing end braces shouldn't panic, but produce an error.

**Comment by simster7** [comment]
As for the `concurrent map read and map write` issue: are you using custom metrics? This recent fix comes to mind: https://github.com/argoproj/argo/pull/3401

**Comment by brabster** [comment]
Thanks @simster7! We're not using custom metrics, I'll see if we're still getting the issue after we've upgraded for those fixes.

**Comment by alexec** [comment]
Fixed and available in v2.10.0-rc5.
